### ISSUE TITLE
Consider changing Should.ThrowAsync to handle synchronous exceptions and cancellation

### ISSUE THREAD

**Issue by paulomorgado** [shouldly/shouldly#471]
Something like this:

```C#
try
{
    await actual();
}
catch (TException ex)
{
    return ex;
}
catch (Exception ex)
{
    throw new ShouldAssertException($"Expected exception: {typeof(TException).FullName}{Environment.NewLine}Actual exception: {ex.GetType().FullName}");
}

throw new ShouldAssertException($"Expected exception: {typeof(TException).FullName} but none was thrown.");
```

in the implementation of [Shloud.ThrowAsync](https://github.com/shouldly/shouldly/blob/master/src/Shouldly/ShouldStaticClasses/ShouldThrowTaskAsync.cs#L47-L85), would allow testing for synchronous operations and operation cancelation.

**Comment by paulomorgado** [comment]
knock, knock!

**Comment by paulomorgado** [comment]
knock, knock!

**Comment by JosephWoodward** [comment]
Sounds like a good idea to me. It should at the very least be able to handle cancellation.

**Comment by paulomorgado** [comment]
Do you want me to send a pull request for it? Where would I need to work on?

**Comment by paulomorgado** [comment]
Do you want me to send a pull request for it? Where would I need to work on?

**Comment by JosephWoodward** [comment]
This has now been fixed as part of #560 and will be in the next beta release.
